colossalai/pipeline/policy/bert.py

```diff
@@ -0,0 +1,326 @@
+            if self.config.add_cross_attention:
+                all_cross_attentions = all_cross_attentions + (layer_outputs[2],)
+
+    if output_hidden_states:
+        all_hidden_states = all_hidden_states + (hidden_states,)
+
+    #end of a stage loop
+    sequence_output = layer_outputs[0] if layer_outputs is not None else None
+
+    if stage_manager.is_last_stage():
+        pooled_output = self.pooler(sequence_output) if self.pooler is not None else None
+        if not return_dict:
+            return (sequence_output, pooled_output) + layer_outputs[1:]
+
+    #output of non-first and non-last stages:
+    if not return_dict:
+        return tuple(v for v in [
+            hidden_states,
+            next_decoder_cache,
+            all_hidden_states,
+            all_self_attentions,
+            all_cross_attentions,
+        ] if v is not None)
+
+    #return dict is not supported at this moment
+    return BaseModelOutputWithPastAndCrossAttentions(
+        last_hidden_state=hidden_states,
+        past_key_values=next_decoder_cache,
+        hidden_states=all_hidden_states,
+        attentions=all_self_attentions,
+        cross_attentions=all_cross_attentions,
+    )
+
+
+# The layer partition policy for bertmodel
+class BertModelPolicy(Policy):
+
+    def __init__(self, stage_manager: PipelineStageManager, num_layers: int, num_stages: int):
+        self.stage_manager = stage_manager
+        self.layers_per_stage = self.distribute_layers(num_layers, num_stages)
+
+    def get_hold_layers(self, module: BertModel) -> List[Module]:
+        """
+        get pipeline layers for current stage
+        """
+        hold_layers = []
+        if self.stage_manager.is_first_stage():
+            hold_layers.append(module.embeddings)
+        num_layers_per_stage_accumulated = self.convert_into_accumulated()
+        hold_layers.extend(module.encoder.layer[num_layers_per_stage_accumulated \
+                    [self.stage_manager.stage-1] if self.stage_manager.stage > 0 else 0:
+                    num_layers_per_stage_accumulated[self.stage_manager.stage]])
+
+        if self.stage_manager.is_last_stage():
+            hold_layers.append(module.pooler)
+
+        return hold_layers
+
+    def get_shared_params(self, module: BertModel) -> List[Dict[int, Tensor]]:
+        '''no shared params in bertmodel'''
+        pass
+
+    def replace_forward(self, module: Module) -> None:
+        module.model.forward = MethodType(partial(bert_model_forward, stage_manager=self.stage_manager), module.model)
+
+    def distribute_layers(self, num, stage_num) -> List[int]:
+        """
+        divide layers into stages
+        """
+        quotient = num // stage_num
+        remainder = num % stage_num
+
+        # calculate the num_layers per stage
+        layers_per_stage = [quotient] * stage_num
+
+        # deal with the rest layers
+        if remainder > 0:
+            middle_stages = (stage_num - 1) // 2
```

Comment:
This can be simplified to one for-loop. `num_stages // 2 - remainder // 2` is the start index to increament `1`.

Reply:
done